File: JavaPartI/le438_findAllAnagramsInAString.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new ArrayList<>();
        int[] dict = new int[26];
        int numele = 0;
        int n = s.length();
        for (char ch : p.toCharArray()) {
            dict[ch - 'a']++;
            if (dict[ch - 'a'] == 1) {
                numele++;
            }
        }
        int k = p.length(); // window size
        for (int i = 0; i < n; i++) {
            char ch = s.charAt(i);
            dict[ch - 'a']--;
            if (i < k) {
                if ( dict[ch - 'a'] == 0) {
                    numele--;
                }
            } else {  
                if ( dict[ch - 'a'] == 0) {
                    numele--;
                } 
                char prev = s.charAt(i - k);
                dict[prev - 'a']++;
                if ( dict[prev - 'a'] == 1) {
                    numele++;
                }
            }
            if (numele == 0) {
                res.add(i - k + 1);
            }
        }
        return res;
    }
}

/// way II 
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new ArrayList<>();
        int[] dict = new int[256];
        int numele = 0;
        int n = s.length();
        for (char ch : p.toCharArray()) {
            dict[ch]++;
            if (dict[ch] == 1) {
                numele++;
            }
        }
        int k = p.length(); // window size
        for (int i = 0; i < n; i++) {
            char ch = s.charAt(i);
            dict[ch]--;
            if ( dict[ch] == 0) {
                numele--;
            }
            if (i >= k) {
                char prev = s.charAt(i - k);
                dict[prev]++;
                if ( dict[prev] == 1) {
                    numele++;
                } 
            }            
            if (numele == 0) {
                res.add(i - k + 1);
            }
        }
        return res;
    }
}

File: pythonPart/le438_findAllAnagramsInAString.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        dictp = {}
        for ch in p:
            if ch in dictp:
                dictp[ch] += 1
            else:
                dictp[ch] = 1
        
        #dictp = Counter(s)]
        n = len(s)
        numele = len(dictp)
        k = len(p)
        res = []
        for i in range(n):
            ch = s[i]
            if ch in dictp:
                dictp[ch] -= 1
                if dictp[ch] == 0:
                    numele -= 1
            if i >= k:
                prev = s[i - k]
                if prev in dictp:
                    dictp[prev] += 1
                    if dictp[prev] == 1:
                        numele += 1
            if numele == 0:
                res.append(i - k + 1)
        return res
